src/main/java/org/opensearch/neuralsearch/query/HybridQueryScorer.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.neuralsearch.query;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+import lombok.Getter;
+
+import org.apache.lucene.search.DisiPriorityQueue;
+import org.apache.lucene.search.DisiWrapper;
+import org.apache.lucene.search.DisjunctionDISIApproximation;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.Scorer;
+import org.apache.lucene.search.Weight;
+
+/**
+ * Class abstracts functionality of Scorer for hybrid query. When iterating over documents in increasing
+ * order of doc id, this class fills up array of scores per sub-query for each doc id. Order in array of scores
+ * corresponds to order of sub-queries in an input Hybrid query.
+ */
+public final class HybridQueryScorer extends Scorer {
+
+    // score for each of sub-query in this hybrid query
+    @Getter
+    private final Scorer[] subScorers;
+
+    private final DisiPriorityQueue subScorersPQ;
+
+    private final DocIdSetIterator docIdIterator;
+
+    final private float[] subScores;
+
+    final private Map<Query, Integer> queryToIndex;
+
+    HybridQueryScorer(Weight weight, Scorer[] subScorers) throws IOException {
```

Comment:
Class is not intended for extension, default scope works as caller and tests are in the same package.

Reply:
final class can have public constructor still. Do you want to block this class from being created outside of same package? Any specific reason?

Reply:
Just by default use most restrictive scope that works, we can change it to public anytime when it's needed. I don't think it will be created anywhere else other than Hybrid Weight class and tests.